membership_prorate/models/account_invoice.py

```diff
@@ -0,0 +1,61 @@
+# Copyright 2015 Tecnativa - Pedro M. Baeza
+# Copyright 2017-18 Tecnativa - David Vidal
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+from odoo import models, api, fields
+from datetime import timedelta
+
+
+class AccountInvoiceLine(models.Model):
+    _inherit = "account.invoice.line"
+
+    def _get_membership_interval(self, product, date):
+        """Get the interval to evaluate as the theoretical membership period.
+        :param product: Product that defines the membership
+        :param date: date object for the requested date (if needed for
+        inherited computations)
+        :return: A tuple with 2 date objects with the beginning and the
+        end of the period
+        """
+        date_from = fields.Date.from_string(product.membership_date_from)
+        date_to = fields.Date.from_string(product.membership_date_to)
+        return date_from, date_to
+
+    def _prepare_invoice_line_prorate_vals(self, invoice_line):
+        product = invoice_line.product_id
+        date_invoice = fields.Date.from_string(
+            invoice_line.invoice_id.date_invoice or fields.Date.today())
+        date_from, date_to = self._get_membership_interval(
+            product, date_invoice)
+        if date_invoice < date_from:
```

Comment:
@chienandalu  here use `If..else...`
```
if date_invoice < date_from:
    date_invoice = date_from
else:
    date_invoice = date_from
```

Reply:
Thanks, @nikul-serpentcs Done

Reply:
Sorry @nikul-serpentcs  , but the change wasn't correct and I revert it.

The purpose of this is to adjust the invoice date inside the membership product period. So the structure is correct:

- The invoice date (e.g.: today) could be **before** that period, thus we would force it to the beginning of that period (the **full qouta** would be applied)
- The invoice date (e.g.: today) could be **after** that period, thus we would force it to the ending of that period (**no qouta** would be applied)
- Or the invoice date could be **inside** that period, , thus we wouldn't change it and the **prorated quota** would be applied.

Reply:
@chienandalu Only Code Review, Not tested in Runbot.
If it's not correct then revert it.
Thanks